app/forms/Choice.scala

```diff
@@ -18,32 +18,40 @@ package forms
 
 import forms.Mapping.requiredRadio
 import play.api.data.{Form, Forms, Mapping}
-import play.api.libs.json.Json
-import utils.validators.forms.FieldValidator.isContainedIn
+import play.api.libs.json.{Format, JsError, JsResult, JsString, JsSuccess, JsValue, Reads}
 
-case class Choice(value: String)
+sealed abstract class Choice(val value: String)
 
 object Choice {
-  implicit val format = Json.format[Choice]
 
   val choiceId = "Choice"
 
-  import AllowedChoiceValues._
-  private val correctChoices = Set(Arrival, Departure, AssociateDUCR, DisassociateDUCR, ShutMucr, Submissions)
-
-  val choiceMapping: Mapping[Choice] = Forms.mapping(
-    "choice" -> requiredRadio("choicePage.input.error.empty")
-      .verifying("choicePage.input.error.incorrectValue", isContainedIn(correctChoices))
-  )(Choice.apply)(Choice.unapply)
+  val choiceMapping: Mapping[Choice] =
+    Forms.mapping("choice" -> requiredRadio("choicePage.input.error.empty"))(Choice.apply)(Choice.unapply)
 
   def form(): Form[Choice] = Form(choiceMapping)
 
-  object AllowedChoiceValues {
-    val Arrival = "EAL"
-    val Departure = "EDL"
-    val AssociateDUCR = "Associate"
-    val DisassociateDUCR = "EAC"
-    val ShutMucr = "CST"
-    val Submissions = "SUB"
+  def unapply(status: Choice): Option[String] = Some(status.value)
+
+  def apply(input: String): Choice =
+    allChoices.find(_.value == input).getOrElse(throw new IllegalArgumentException("Incorrect choice"))
+
+  implicit object ChoiceValueFormat extends Format[Choice] {
```

Comment:
Much, much better than Strings.  But it still irks me that we need this much code to convert a string to a 'value'.  Roll-on proper enum support in scala!  :-)

Reply:
There is very good support but in third party library - https://github.com/lloydmeta/enumeratum - But here we hit the problem of type system expressiveness and type class derivation.